File: areAnagram.js
function areAnagram(str1, str2) {
    // Step 1: If lengths are different, they can't be anagrams
    if (str1.length !== str2.length) {
        return false;
    }

    // Step 2: Create a character count map for str1
    const charCount = {};

    for (let char of str1) {
        // If character already exists, increment count
        // Otherwise, initialize it to 1
        charCount[char] = (charCount[char] || 0) + 1;
    }

    // Step 3: Subtract character counts using str2
    for (let char of str2) {
        // If character doesn't exist or count is zero, not an anagram
        if (!charCount[char]) {
            return false;
        }

        // Decrease the count for matched character
        charCount[char]--;
    }

    // Step 4: All characters matched correctly
    return true;
}

str1="test";
str2="stay";

result = areAnagram(str1,str2);
if (result){
    console.log("given strings are Anagram");
}
else{
    console.log("Not Anagram");
}

File: areAnagram.py
def are_anagram(str1, str2):
    if len(str1) != len(str2):
        return False

    count = {}

    # ✅ First loop: build frequency map from str1
    for char in str1:
        count[char] = count.get(char, 0) + 1

    # ✅ Second loop: validate using str2
    for char in str2:
        if char not in count or count[char] == 0:
            return False
        count[char] -= 1

    return True

str1 = input("Enter the first string: ")
str2 = input("Enter the second string: ")

if are_anagram(str1, str2):
    print("The strings are anagrams")
else:
    print("The strings are not anagrams")
